### PR TITLE
Pangolin Roll

### PR DESCRIPTION
**Pull request by 0xkobe**

### CODE REVIEW

File: contracts/pangolin-roll/PangolinRoll.sol [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+import "./libraries/SafeERC20.sol";
+import "./libraries/Ownable.sol";
+
+// PangolinRoll helps your migrate your existing Sushiswap LP tokens to Pangolin LP ones
+contract PangolinRoll is Ownable {
+    using SafeERC20 for IERC20;
+
+    IPangolinRouter public oldRouter;
+    IPangolinRouter public router;
+    IERC20 public pngToken = IERC20(0x60781C2586D68229fde47564546784ab3fACA982); // png token address
```
Comment (bmino) [inline contracts/pangolin-roll/PangolinRoll.sol:19]: let's have this passed in as a constructor argument to allow this on other networks and for ease of testing. these variables can be `immutable` to accommodate this

File: contracts/pangolin-roll/PangolinRoll.sol [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        // Send remaining tokens to msg.sender
+        if (amountA > pooledAmountA) {
+            IERC20(tokenA).safeTransfer(msg.sender, amountA - pooledAmountA);
+        }
+        if (amountB > pooledAmountB) {
+            IERC20(tokenB).safeTransfer(msg.sender, amountB - pooledAmountB);
+        }
+
+        // Transfer user a single PNG token if there are any remaining and user has not received one yet
+        if (address(pngToken) != address(0)) {
```
[the remark this answers is not in the document]
Comment (0xkobe) [reply contracts/pangolin-roll/PangolinRoll.sol:80]: If we put pngToken in constructor, this check is needed, I guess.

File: contracts/pangolin-roll/PangolinRoll.sol [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+            uint256 pngSupply = pngToken.balanceOf(address(this));
+            if (pngSupply > 0 && !checkList[msg.sender]) {
```
Comment (bmino) [inline contracts/pangolin-roll/PangolinRoll.sol:88]: This should check if the supply is greater than 1e18 or we risk bricking the contract if anybody sends PNG to it where the balance is between [0, 1]

File: contracts/pangolin-roll/PangolinRoll.sol (hunk 2 above)
[the remark this answers is not in the document]
Comment (bmino) [reply contracts/pangolin-roll/PangolinRoll.sol:86]: do the require in the constructor once so we don't spend gas on every migrate checking

File: contracts/pangolin-roll/PangolinRoll.sol [hunk 4]
```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.6.0;
```
Comment (bmino) [inline contracts/pangolin-roll/PangolinRoll.sol:3]: we can certainly use solidity 0.8.0+ and interact with other contracts that are earlier versions. hit me up and we can work through how dependencies for both versions can be managed in the repo

**Pull request closed without merging**